t/t2291-job-update-queue.t

```diff
@@ -0,0 +1,152 @@
+#!/bin/sh
+test_description='Test update of job queue'
+
+. $(dirname $0)/sharness.sh
+
+if flux job submit --help 2>&1 | grep -q sign-type; then
+	test_set_prereq HAVE_FLUX_SECURITY
+fi
+
+test_under_flux 4 full
+
+flux setattr log-stderr-level 1
+
+runas_guest() {
+	local userid=$(($(id -u)+1))
+        FLUX_HANDLE_USERID=$userid FLUX_HANDLE_ROLEMASK=0x2 "$@"
+}
+
+submit_held_job_as_guest()
+{
+	local duration=$1
+	local userid=$(($(id -u)+1))
+        flux run --dry-run -t $duration true | \
+          flux python ${SHARNESS_TEST_SRCDIR}/scripts/sign-as.py $userid \
+            >job.signed
+        FLUX_HANDLE_USERID=$userid \
+          flux job submit --flags=signed --urgency=0 job.signed
+}
+
+test_expect_success 'config queues and resources' '
+	flux R encode -r 0-3 -p batch:0-3 -p debug:3 \
+	   | tr -d "\n" \
+	   | flux kvs put -r resource.R=- &&
+	flux config load <<-EOT &&
+	[queues.batch]
+	requires = [ "batch" ]
+	policy.limits.duration = "1h"
+	policy.jobspec.defaults.system.duration = "1h"
+
+	[queues.debug]
+	requires = [ "debug" ]
+	policy.limits.duration = "1m"
+	policy.jobspec.defaults.system.duration = "1m"
+
+	[policy.jobspec.defaults.system]
+	queue = "batch"
+	EOT
+	flux queue start --all &&
+	flux module unload sched-simple &&
+	flux module reload resource &&
+	flux module load sched-simple &&
+	flux queue list &&
+	flux resource list -o rlist
+'
+test_expect_success 'invalid queue update RPC fails' '
+	jobid=$(flux submit --urgency=hold -n1 hostname) &&
+	echo "{\"id\": $(flux job id $jobid),\
+	       \"updates\": {\"attributes.system.queue\": 42}\
+	      }" \
+            | ${FLUX_BUILD_DIR}/t/request/rpc job-manager.update 22 && # EINVAL
+	flux cancel $jobid &&
+	flux job wait-event $jobid clean
+'
+test_expect_success 'update of invalid job fails' '
+	test_must_fail flux update f1234 queue=batch
+'
+test_expect_success 'update queue of running job fails' '
+	jobid=$(flux submit --wait-event=start -n1 sleep 300) &&
+	test_must_fail flux update $jobid queue=debug 2>running.err &&
+	flux cancel $jobid &&
+	flux job wait-event $jobid clean &&
+	grep "update of queue for running job not supported" running.err
+'
+test_expect_success 'update to invalid queue fails' '
+	jobid=$(flux submit -q debug --urgency=hold hostname) &&
+	test_must_fail flux update $jobid queue=foo
+'
+test_expect_success 'update to same queue fails' '
+	test_must_fail flux update $jobid queue=debug
+'
+test_expect_success 'update to batch queue works' '
+	flux update $jobid queue=batch &&
+	test_debug "flux job eventlog $jobid" &&
+	flux job eventlog $jobid \
+	  | grep jobspec-update \
+	  | grep attributes.system.queue=\"batch\"
+'
```

Comment:
haven't tried yet, but could test `flux job list` to see if `job-list` got update (here and perhaps few other places)?  I think it should work ... if it doesn't, can be a TODO item for me to minimally add some tests in t2260-job-list.

Note, adding here could be annoying as you gotta spin a bit to wait for job-list eventual consistency.

Reply:
Yeah, I didn't do it because of the raciness, and job update processing in job-list is tested elsewhere so it could be considered redundant here.